lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,182 @@
+}
+
+// InitFileSharedWriter wraps the provided [os.File] in a writer that is thread safe,
+// with ability to enable filesystem notification watch and reopen file on specific events.
+func InitFileSharedWriter(logFile *os.File, flag int, mode fs.FileMode) (io.Writer, error) {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if fileSharedWriter != nil {
+		return nil, trace.BadParameter("file shared writer already initialized")
+	}
+	fileSharedWriter = &FileSharedWriter{fileFlag: flag, fileMode: mode}
+	fileSharedWriter.logFile.Store(logFile)
+
+	return fileSharedWriter, nil
+}
+
+// CloseFileSharedWriter closes the file shared writer and frees up resources.
+func CloseFileSharedWriter() {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	if fileSharedWriter != nil {
+		if err := fileSharedWriter.Close(); err != nil {
+			slog.ErrorContext(context.Background(), "Failed to close file shared writer", "error", err)
+		}
+		fileSharedWriter = nil
+	}
+}
+
+// GetFileSharedWriter returns instance of the file shared writer.
+func GetFileSharedWriter() *FileSharedWriter {
+	initLock.Lock()
+	defer initLock.Unlock()
+
+	return fileSharedWriter
+}
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.logFile.Load().Write(b)
+}
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	logFile, err := os.OpenFile(s.logFile.Load().Name(), s.fileFlag, s.fileMode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	prevLogFile := s.logFile.Swap(logFile)
+	if err := prevLogFile.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// Close stops watcher loop and frees resources.
+func (s *FileSharedWriter) Close() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if watcher := s.watcher.Swap(nil); watcher != nil {
+		return trace.Wrap(watcher.Close())
+	}
+	s.logFile.Swap(nil)
+
+	return nil
+}
+
+// RunWatcherReopen runs a filesystem watcher for rename/remove events to reopen the log.
+func (s *FileSharedWriter) RunWatcherReopen(ctx context.Context) error {
+	return s.runWatcherFunc(ctx, s.Reopen)
+}
+
+// runWatcherFunc spawns goroutine with the watcher loop to consume events of renaming
+// or removing the log file to trigger the action function when event appeared.
+func (s *FileSharedWriter) runWatcherFunc(ctx context.Context, action func() error) error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	context.AfterFunc(ctx, func() {
+		if err := watcher.Close(); err != nil {
+			slog.ErrorContext(context.Background(), "Failed to close file watcher", "error", err)
+		}
+	})
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				if s.logFile.Load().Name() == event.Name && (event.Has(fsnotify.Rename) || event.Has(fsnotify.Remove)) {
```

Comment:
Should we recreate the file if it's not there when we try to reopen it, or should we only reopen the file once something else puts a new file in its place?

Reply:
we use `os.O_CREATE` flag to re-open file, so we should create it always